Starting the rehabilitation of the array episode. (#506)
carpentries-incubator/hpc-intro: episodes/21-array.Rmd

Reviewer: ```suggestion
One common approach to parallel computing is to distribute a computation across multiple processes. These processes cooperate by exchanging information during execution using Message Passing Interface (MPI).
```

Reviewer: This instructor note is now obsolete and can be removed.

:::::::::::::::::::::::::::::::::::::::: questions

- What are job arrays?
- What benefits do job arrays bring?
- What type of jobs would benefit from job arrays?

::::::::::::::::::::::::::::::::::::::::::::::::::

> Parallel computing is a technique used to divide big tasks into smaller ones

Reviewer: The statement is narrowing down parallel computing to *task decomposition* only.

This could be made clearer as:

```suggestion
Parallel computing allows multiple computational tasks to execute simultaneously in order to reduce execution time or increase throughput. Depending on the application, the work may be divided into cooperating subtasks that communicate with one another, or into completely independent tasks that execute separately.
```

Author: Better wording, thanks! I've incorporated most of these suggestions, with a few additional tweaks.

:::::::::::::::::::::::::::::::::::::::: questions

- What are job arrays?
- What benefits do job arrays bring?
- What type of jobs would benefit from job arrays?

::::::::::::::::::::::::::::::::::::::::::::::::::

Parallel computing allows multiple computational tasks to execute
simultaneously in order to reduce execution time for a task, or
increase throughput for multiple tasks. Depending on the application,

> the work may be divided into cooperating subtasks that communicate with

Reviewer: for reconnecting the next sentences (same terminology):

```suggestion
the workload may be divided into cooperating subtasks that communicate with
```

one another, or into independent tasks that execute separately.

One common approach to parallel computing is to distribute computation across

> multiple processes, and having these processes cooperate by exchanging

Reviewer: The sentence ", and having these processes" is breaking the sentence flow.

```suggestion
multiple processes that cooperate by exchanging
```
filler clears this

> information during execution using the Message Passing Inteface. This

Reviewer: typo

```suggestion
information during execution using the Message Passing Interface. This
```

> is the "MPI" that we saw as a dependency of the `am

Reviewer: incomplete sentence

```suggestion
This is the "MPI" that we saw as a dependency of the application `amdahl`.
```

Author: Fixed, not sure what happened here.

Not all workloads require processes to cooperate. Many scientific workflows
are made up of independent serial jobs. For these cases, `r config$sched$name`
provides Job Arrays, allowing many similar jobs to be submitted and managed together.

For instance,
you might need to run the same task on several independent input
files, or you may have multiple serial tasks that take some parameter,
and you need to explore several values of the parameter. These

> tasks are also sometimes called "high-throughput" computing.

Reviewer: distinguishing between "tasks" and "workflows".

```suggestion
These workflows are commonly referred to as "high-throughput computing" (HTC).
```

Operationally speaking, a Job Array is a collection of related batch
jobs submitted using a single job script.

:::::::::::::::::::::::::::::::::::::::  challenge

What distinguishes workloads that are suitable for job arrays from those
that require traditional parallel programming?
Name some examples.

:::::::::::::::  solution

> Tasks appropriate for array jobs are "high-throughput", where

Reviewer: restructuring the sentence for clarity:

```suggestion
Array jobs are well suited to "high-throughput workloads", where the same operation is performed many times, repeatedly with different inputs or parameter values, and each task is independent of the others.
```

Author: Obsoleted by other changes.

the same thing needs to be done many times, possibly over a set
of parameters, but where each task in dependent of the others.

For example, running the same statistical analysis on a large
number of independent input files is a good candidate for an
array solution.

Parallel tasks which have interactions between the various

> parallel threads need to communicate between tasks at

Reviewer: instead of "parallel threads", in order to make this clear from shared-memory threading:

```suggestion
parallel processes/tasks
```

Author: In this context, "task" is overloaded, I've hopefully clarified that the "task" is the larger scope which includes multiple "processes".  Wording changes are a bit different from what was suggested, but meet the goal, I think.

Reviewer: The concept is intact, we can make it short and clean:

```suggestion
Parallel tasks that require communication between processes during execution are not appropriate for job arrays.
```

run-time, and are not appropriate for job arrays.

For example, most parallel scientific codes that run in
parallel have a requirement to communicate between parallel
elements at run-time, and are not appropriate for job arrays.

Similarly, serial tasks which only need to be run once do not
benefit from parallelism. Aggregating unrelated tasks into an
array merely for the sake of grouping does not make sense.

:::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::::::::::::

> We have aleady seen how to submit a single seriial job to the

Reviewer: typo:

```suggestion
already
```

> We have aleady seen how to submit a single seriial job to the

Reviewer: typo:

```suggestion
serial
```

> Slurm scheduler.

Reviewer: minor instead of using "Slurm" directly:
```suggestion
``r config$sched$name``
```

:::::::::::::::::::::::::::::::::::::::: questions

- What are job arrays?
- What benefits do job arrays bring?
- What type of jobs would benefit from job arrays?

::::::::::::::::::::::::::::::::::::::::::::::::::

Parallel computing allows multiple computational tasks to execute
simultaneously in order to reduce execution time for a task, or
increase throughput for multiple tasks. Depending on the application,
the workload may be divided into cooperating subtasks that communicate with
one another, or into independent tasks that execute separately.

One common approach to parallel computing is to distribute computation across
multiple processes that cooperate by exchanging information during
execution using the Message Passing Interface. This is the "MPI" that we
saw as a dependency of the `amdahl` application.

Not all workloads require processes to cooperate. Many scientific workflows

> are made up of independent serial jobs. For these cases, `r config$sched$name`

Reviewer 2: Misleading statement: JobArrays are not at all constrained to serial executables. They are often a way to nest a task-parallel code within a data-parallel workflow.

```suggestion
are made up of independent jobs. For these cases, `r config$sched$name`
```

Author: Accetped.

provides Job Arrays, allowing many similar jobs to be submitted and managed together.

For instance,
you might need to run the same task on several independent input
files, or you may have multiple serial tasks that take some parameter,
and you need to explore several values of the parameter.
Workflows made up of these independent elements are 
also sometimes called "high-throughput" computing.

Operationally speaking, a Job Array is a collection of related batch
jobs submitted using a single job script.

:::::::::::::::::::::::::::::::::::::::  challenge

What distinguishes workloads that are suitable for job arrays from those
that require traditional parallel programming?

> Name some examples.

Reviewer 2: ```suggestion
Describe some examples.
```

:::::::::::::::  solution

Tasks appropriate for array jobs are "high-throughput", where
the same thing needs to be done many times, possibly over a set
of parameters, but where each task is independent of the others.

For example, running the same statistical analysis on a large
number of independent input files is a good candidate for an
array solution.

Parallel tasks which have interactions between the various
parallel processes need to communicate between processes at
run-time, and are not appropriate for job arrays.

For example, most parallel scientific codes that run in
parallel have a requirement to communicate between parallel
elements at run-time, and are not appropriate for job arrays.

Similarly, serial tasks which only need to be run once do not
benefit from parallelism. Aggregating unrelated tasks into an

> array merely for the sake of grouping does not make sense.

Reviewer 2: I like this clarification!

Author: Yay!

multiple processors. This is usually accomplished by using the Message Passage
Interface (MPI) which is a standardised way for CPU cores to communicate with
one another while working together on a task. Software has to be written
specifically to utilize MPI to take advantage of this. We will see more
of this in a later episode.

> Another, simpler form of parallel computing is an array job. This type of

Reviewer: ```suggestion
Not all workloads require processes to communicate. Many scientific workflows consists of a large number of independent serial jobs. For these cases, `r config$sched$name` provides Job Arrays, allowing many similar jobs to be submitted and managed together.

> Another, simpler form of parallel computing is an array job. This type of

Reviewer: ```suggestion
A Job Array is a collection of related batch jobs submitted using a single job script.
```

Reviewer: a *Job Array* before explanation.

job is advantageous if the same software has to be run several times
in a way where the separate runs do not share data, and do not need
to run concurrently. For instance,
you might need to run the same task on several independent input
files, or you may have multiple serial tasks that take some parameter,
and you need to explore several values of the parameter. These
tasks are also sometimes called "high-throughput" computing.

:::::::::::::::::::::::::::::::::::::::  challenge

> What is the important distinguishing feature between tasks

Reviewer: ```suggestion
What distinguishes workloads that are suitable for job arrays from those that require traditional parallel programming?
```

that are appropriate for array jobs and task that are not?
Name some examples.

:::::::::::::::  solution

> Tasks appropriate for array jobs are "high-throughput", where

Reviewer: ```suggestion
Array jobs are well suited to "high-throughput workloads", where the same operation is performed many times, possibly with different input parameters, and each task is independent of the others.
```

the same thing needs to be done many times, possibly over a set
of parameters, but where each task in dependent of the others.

For example, running the same statistical analysis on a large
number of independent input files is a good candidate for an
array solution.

Parallel tasks which have interactions between the various
parallel threads need to communicate between tasks at
run-time, and are not appropriate for job arrays.

For example, most parallel scientific codes that run in
parallel have a requirement to communicate between parallel
elements at run-time, and are not appropriate for job arrays.

Similarly, serial tasks which only need to be run once do not
benefit from parallelism. Aggregating unrelated tasks into an
array merely for the sake of grouping does not make sense.

:::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::::

> We have aleady seen how to submit a single serial job to the

Reviewer: typo

```suggestion
already
```

Slurm scheduler, and also a single parallel jobs.

Array jobs differ in character, but are specified in a similar manner.
The batch file describes the array to the system, and at submission
time, all of the array elements are submitted together. They may
dispatch at different times, and may or may not run concurrently.

Within

> the execution envioronment of an array element, a number of environment

Reviewer: typo
```suggestion
environment
```
